scenes/performance.gd

```diff
@@ -151,3 +169,107 @@ func _exit_tree():
+	return ret
+
+func reload_network_users():
+	var users := get_all_network_users()
+	for u in users:
+		if !ingame_users.has(u):
+			_add_new_user(u)
+			
+
+func _add_new_user(u) -> IngameUser:
+	var nu = IngameUser.new()
+	ingame_users[u] = nu
+	ingame_scores_viewer.reload_users(ingame_users)
+	return nu
+
+func _on_user_connected(id : int):
+	reload_network_users()
+	if is_multiplayer_authority():
+		var uds := {}
+		for u in ingame_users:
+			uds[u] = u.get_data()
+		rpc("_set_all_user_data", uds)
+
+func _on_user_disconnected(id : int):
+	reload_network_users()
+
+func _on_server_disconnected():
+	# TODO
+	pass
+
+func _on_connected_to_server():
+	ingame_users["%s" % multiplayer.get_unique_id()] = self_user
+	reload_network_users()
+
+func update_user_data(uid : String):
+	# user data for has changed.
+	if multiplayer.has_multiplayer_peer():
+		await GBackend.await_connection()
+		if multiplayer.multiplayer_peer.get_connection_status() == MultiplayerPeer.CONNECTION_CONNECTED:
+			rpc("_set_user_data", self_user.get_data())
+
+func broadcast_all_user_data():
+	if multiplayer.has_multiplayer_peer():
+		await GBackend.await_connection()
+		if multiplayer.multiplayer_peer.get_connection_status() == MultiplayerPeer.CONNECTION_CONNECTED:
+			print_debug("broadcasting all user data")
+			var uds := {}
+			for u in ingame_users:
+				uds[u] = u.get_data()
+			rpc("_set_all_user_data", uds)
+
+@rpc("authority", "call_remote", "reliable") func _set_all_user_data(data):
+	print("rpc _set_all_user_data called with {%s}" % data)
+	for k in data:
+		if !(k is String):
+			# invalid key
+			push_error("invalid user key %s" % k)
+			continue
+		var ig : IngameUser
+		if ingame_users.has(k):
+			ig = ingame_users[k]
+		elif k == ("%s" % multiplayer.get_unique_id()):
+			ingame_users[k] = self_user
+			ig = self_user
+		else:
+			ig = _add_new_user(k)
+			
+		var udata = data[k]
+		if !(udata is Dictionary):
+			# invalid user data
+			push_error("invalid user data %s" % udata)
+			continue
+		ig.parse_data(udata)
+
+@rpc("any_peer", "call_remote", "reliable") func _set_user_data(data):
+	print("received remote _set_user_data call with [%s]" % data)
+	var uid : String = "%d" % multiplayer.get_remote_sender_id()
+	if !(data is Dictionary):
+		push_error("invalid user data for uid [%s]" % uid)
+		return
+	if !ingame_users.has(uid):
+		push_warning("something went wrong, non existing user is sending data")
+		return
+	ingame_users[uid].parse_data(data)
+
+func _on_connect_instrument() -> void:
+	print_debug(performance_instrument)
+	performance_instrument.Notes.good_note_started.connect(_on_good_note_started)
+
+func _on_good_note_started(note_index : int, time_error : float) -> void:
+	self_user._on_good_note(note_index, time_error)
+	if multiplayer.has_multiplayer_peer():
+		await GBackend.await_connection()
+		if multiplayer.multiplayer_peer.get_connection_status() == MultiplayerPeer.CONNECTION_CONNECTED:
+			rpc("_set_user_data", self_user.get_data())
```

Comment:
Add double white lines between functions. Refer to the style guide for exact spacing rules, there are a few more cases where a double line is required.